### PR TITLE
[None][feat] Make the should_use_spec_decode logic a bit smarter

### PR DESCRIPTION
**Pull request by zheyuf**
## Description
Previously, we disabled speculation solely when len(active_requests) > max_concurrency, which was conservative. We now gate by a better estimated per-step capacity (min(len(active_requests), max_batch_size, max_num_tokens // (1 + draft_len))), so speculation isn’t unnecessarily turned off when the schedulable batch is small. This respects real decoding limits, improving GPU utilization and latency. 

## Test Coverage
There are existing unit tests in TensorRTLLM/tests/unittest/_torch/speculative/test_dynamic_spec_decode.py. Also added  some quick unit tests in this file.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Summary by CodeRabbit

* **New Features**
  * Speculative decoding now factors batch size, overall token budget, and draft length when deciding to enable speculation, improving efficiency across workloads.

* **Refactor**
  * Decision gate changed from a pure concurrency check to a token-budget- and batch-aware approach for more predictable and constrained speculation behavior.

* **Tests**
  * Unit tests updated and a new test added to cover varied scenarios and edge cases for the expanded decision criteria.

### CODE REVIEW

**Review by mikeiovine — APPROVED** [review]

File: tests/unittest/_torch/speculative/test_dynamic_spec_decode.py [hunk 1]
```diff
@@ -87,5 +88,62 @@ def mock_should_use_spec_decode(self, requests):
...
+    drafter = _DummyDrafter(max_concurrency=6)
+
+    # Compare min(len(requests), max_batch_size, token_cap) with max_concurrency
+
+    # Small active_requests ON case: num_effective_requests = min(5, 8, very_large) = 5 <= 6 → True
+    active_requests = [object()] * 5
+    assert drafter.should_use_spec_decode(active_requests,
+                                          max_batch_size=8,
+                                          max_num_tokens=4096 * 8,
+                                          max_draft_len=4) is True
```
[the remark this answers is not in the document]
Comment (zheyuf) [reply tests/unittest/_torch/speculative/test_dynamic_spec_decode.py]: Thanks. Updated the code.

File: tensorrt_llm/_torch/speculative/drafter.py [hunk 2]
```diff
@@ -27,12 +27,28 @@ def prepare_draft_tokens(
...
+        # Inputs typically validated upstream: max_batch_size>0, max_num_tokens>0, max_draft_len>=0
+
+        if self.max_concurrency is None:
+            return True
+
+        # Defensive guards; keep behavior explicit for zero/empty cases
+        if not requests or max_batch_size <= 0 or max_num_tokens <= 0:
+            return False
+
+        tokens_per_request = 1 + max(0, max_draft_len)
```
Comment (yweng0828) [inline tensorrt_llm/_torch/speculative/drafter.py:48]: Hi, I think the `max_draft_len` will always be greater than or equal to 0, is that right? And the default `max_draft_len` should be `0`. So I wonder if `max(0, max_draft_len)` can be simplified to `max_draft_len`.

**Review by yweng0828 — APPROVED** [review]

**Review by mikeiovine — APPROVED** [review]
